rostopic: handling multiple array index specifications

rhaschke opened the pull request:
Please consider this patch that fixes an issue with multiple array index specifications for subfields in rostopic echo. E.g.
- `rostopic echo /tactile_mid_prox/sensors[1]/proximal[1]` failed with

> Topic name 'sensors[1]/proximal' contains '[' but does not end with ']'
- `rostopic echo /tactile_mid_prox/sensors[1]/proximal[1]` failed with

> Topic name 'sensors[1]/proximal[1]' contains invalid slice argument '1]/proximal[1': non-integer array index step '1]/proximal[1'

The reason were two issues:
- First, in function `msgevalgen`, it was assumed that there will be only a single index specification. I fixed, that by a) looking for the matching closing bracket (instead of using the whole rest of the string) and b) calling `msgevalgen` recursively on trailing sub fields.
- Second, in function `_rostopic_echo` when extracting the `type_information` for the to-be-displayed values, a second index specification was not expected too, resulting in an attempt to extract the type for `subtype[]` instead of `subtype`. This was easily fixed, dropping the brackets.

rhaschke committed:
rostopic: handling multiple array index specifications

ros-pull-request-builder commented:
Can one of the admins verify this patch?

rhaschke committed:
gracefully handle IndexErrors when accessing message fields

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -300,28 +300,37 @@ def msgevalgen(pattern):
+    rec_msgeval = None
+    index_start = msg_attribute.find('[')
+    if index_start != -1:
+        index_end = msg_attribute.find(']')
```
Shouldn't the find search start from `index_start`? Or at least check the order of the start and end indices afterwards?

rhaschke replied:
You are right. One should better use `msg_attribute.find(']', index_start)'.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -300,28 +300,37 @@ def msgevalgen(pattern):
             value = _get_nested_attribute(msg, msg_attribute)
         except AttributeError as e:
-            sys.stdout.write("no field named [%s]"%pattern+"\n")
+            if len(msg) > 1:
```
Isn't this happening if you try to output e.g. `/tactile_mid_prox/sensors[1]/proximal`? Or if the message field is a string?

rhaschke replied:
This addition is related to multiple slicings like /topic/field[0:3]/field[0:2]. In this case it's neither clear how to output the product of array results from these multiply slicings, nor can _get_nested_attribute() handle lists of msgs.
Hence, the warning about multiple slicings. The standard warning would show `no field named [field[0:2]]`, which doesn't explain the issue for the user.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -300,28 +300,37 @@ def msgevalgen(pattern):
         except AttributeError as e:
-            sys.stdout.write("no field named [%s]"%pattern+"\n")
+            if len(msg) > 1:
+                sys.stdout.write("slicing only supported for last array index\n")
```
Should this go to `stderr` instead? Same two lines below.

rhaschke replied:
I simply kept the stdout that was used before. If you want to change it, go ahead.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -788,7 +797,9 @@ def _rostopic_echo(topic, callback_echo, bag_file=None, echo_all_topics=False):
                     if fields:
-                        submsg_class = roslib.message.get_message_class(type_information)
+                        submsg_class = roslib.message.get_message_class(type_information.split('[')[0])
+                        if not submsg_class:
```
Why is this necessary? Can you please provide an example where this is the case.

rhaschke replied:
`type_information` for an array element will contain the array subscript, e.g. `field[2]`. 
However, this type string doesn't have a valid message class. The type is named `field` in this case. Hence, we need to drop the brackets.

dirk-thomas commented:
Since the complexity of this message evaluation gets more and more complex to support all these cases can you please add unit tests for this new feature.

rhaschke committed:
integrated comments from dirk-thomas and added more documentation

rhaschke commented:
I guess, in order to have unit tests for these case, I will have to launch a talker that publishes a nested array message. Do you suggest to add such a msg definition for testing only? Or can you suggest an existing message type using nested arrays?
Where should this talker code go to? Currently the talker node used for testing comes from package rospy.

rhaschke committed:
reworked rostopic.msgevalgen() to allow for multiple slicings

- adapted simpler and more efficient code from rqt_plot
- allow for multiple slicings calling _msgeval recursively
  return value of applying the function are nested lists in this case

rhaschke committed:
added slicing unit tests for rostopic.msgevalgen()

rhaschke committed:
send error output to stderr, not stdout

rhaschke commented:
@dirk-thomas 
As I didn't get feedback on my question how and where to add nested arrayed messages for testing,
I have chosen another approach for unit tests: offline testing of rostopic.msgevalgen() using existing msg types `ArrayVal` and `Val` (i.e. a list of strings) from `test_rospy.msg`.

To this end, **I had to re-enable nosetests in cmake**. However, the **other tests** already existing in test_rostopic_unit.py **failed. Please decide how to proceed with these tests** (fix or disable).

Further, I reworked rostopic.msgevalgen() following a similar pattern as in rqt_plot (as was already suggested by a [TODO](https://github.com/ros/ros_comm/blob/indigo-devel/tools/rostopic/src/rostopic/__init__.py#L288). However, I allow for multiple slicings now, resulting in nested arrays as output (see [unit tests](https://github.com/ubi-agni/ros_comm/blob/indigo-devel/tools/rostopic/test/test_rostopic_unit.py#L428) for examples).

Finally, I fixed an unrelated typo: Probably [valid_type should be named valid_type_s_.](https://github.com/ros/ros_comm/blob/indigo-devel/tools/rostopic/src/rostopic/__init__.py#L263)

rhaschke commented:
@ros-pull-request-builder

dirk-thomas commented:
Please don't add unrelated changes into the same pull request but create separate ones. It makes it not only more difficult to review but also prevent some changes from being merged early while waiting for others to be iterated on.

I have committed the following changes separately:
- using the `queue_size` argument to avoid a warning: 05ad10f1750260a9f6f9e31a6c7842b7fbf511ac
- fix the string type check: 13c44d17f5725809686ae1457f072e850ee7a305
- reenable the offline test: 3c2a9d9b4e45451c301e03855e1fd84479dd5f4a
- moved the remaining disabled unit test to a separate test package where it can create the necessary test messages and was reenabled again: bd10f702dcf4c447cc421ed3fd4838f9173afe28

Please rebase the PR. The newly added test code should go into https://github.com/ros/ros_comm/blob/bd10f702dcf4c447cc421ed3fd4838f9173afe28/test/test_rostopic/test/test_rostopic_unit.py

rhaschke commented:
@dirk-thomas 
I rebased the PR as requested.

dirk-thomas commented on test/test_rostopic/test/test_rostopic_unit.py:
```diff
@@ -388,3 +388,45 @@ def test_create_field_filter(self):
         self.assertEquals(['seq', 'stamp'], list(f(m)))
+
+    def test_slicing(self):
+        from test_rospy.msg import ArrayVal, Val
```
The `test_rostopic` package does not depend on `test_rospy`. Please add the necessary messages to `test_rostopic` and use them from there.

dirk-thomas commented on test/test_rostopic/test/test_rostopic_unit.py:
```diff
@@ -388,3 +388,45 @@ def test_create_field_filter(self):
+
+        self.assertEqual(f(''), None)
+        self.assertEqual(f('/'), None)
+        self.assertListEqual(f('/vals')(msg),    msg.vals)
```
Please remove the extra white spaces following PEP8.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -285,45 +285,59 @@ def _rostopic_bw(topic, window_size=-1):
+    fields = [f for f in pattern.split('/') if f]
+    for f in fields:
+        if '[' in f:
+            field_name, rest = f.split('[')
```
The `split` operation should be limited to return only 2 values using the `maxsplit` argument.

rhaschke replied:
Sure. Didn't know about this arg. Always learning new stuff ;-)

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -285,45 +285,59 @@ def _rostopic_bw(topic, window_size=-1):
+    for f in fields:
+        if '[' in f:
+            field_name, rest = f.split('[')
+            rest = rest[:rest.find(']')]
```
Since we don't allow `field[1][2]` here the closing bracket should always be the last character, no?

rhaschke replied:
Yes. Good point.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -285,45 +285,59 @@ def _rostopic_bw(topic, window_size=-1):
+
+    def _msgeval(msg, evals):
+        if not evals:
+            return msg
```
Could this conditional be removed? It seems to be already covered by the conditional return line of the `msgevalgen` function.

rhaschke replied:
Indeed. Cross-checked with additional unittest, checking for /field/ (slash in the end).

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -285,45 +285,59 @@ def _rostopic_bw(topic, window_size=-1):
+            try: # access field first
+                msg = getattr(msg, field_name)
+            except AttributeError:
+                sys.stderr.write("no field named [%s]\n" % field_name)
```
May be more context information would be helpful for users. E.g. show the full "path" to the unresolvable field.

Same at the next except block.

rhaschke replied:
added full pattern

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -285,45 +285,59 @@ def _rostopic_bw(topic, window_size=-1):
+                # we need to recursively call msg_eval() with the rest of evals
+                # in order to handle nested slices
+                if isinstance(msg, list):
+                    rest = evals[i+1:]
```
Please add spaces around the `+` following PEP8.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -788,7 +802,9 @@ def _rostopic_echo(topic, callback_echo, bag_file=None, echo_all_topics=False):
-                        submsg_class = roslib.message.get_message_class(type_information)
+                        submsg_class = roslib.message.get_message_class(type_information.split('[')[0])
+                        if not submsg_class:
+                            raise ROSTopicException("Cannot load message class for [%s]. Are your messages built?"%type_information)
```
Please add spaces around the`%` following PEP8.

dirk-thomas replied:
Can you please address this comment too.

rhaschke replied:
Sorry, missed that one.

dirk-thomas commented:
Besides the few small comments the algorithm looks solid. This is a great improvement. Thanks.

rhaschke committed:
removed dependency on test_rospy.msg, copying the required messages to test_rostopic.msg

rhaschke committed:
PEP8 corrections

rhaschke committed:
replaced sys.stderr.write() with print()

to correctly auto-include newline

rhaschke committed:
adaptions suggested by dirk-thomas

rhaschke commented:
@dirk-thomas I fixed the small issues you listed. Thanks a lot for your comments.

dirk-thomas commented on test/test_rostopic/CMakeLists.txt:
```diff
@@ -12,6 +12,8 @@ if(CATKIN_ENABLE_TESTING)
     Simple.msg
     TVals.msg
+    Val.msg
+    ArrayVal.msg
```
Please add the new messages in alphabetical order.

rhaschke replied:
OK. Fixed.

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -280,50 +280,62 @@ def _rostopic_bw(topic, window_size=-1):
-            value = value[array_index_or_slice_object]
-        return value
-    return msgeval
+        return _msgeval(msg, evals)
```
Since `msgeval` is now basically empty you could move the implementation of `_msgeval` here and get rid of `_msgeval`.

dirk-thomas replied:
Can you please address this comment too.

rhaschke replied:
No, it cannot be removed easily. _msgeval() is called recursively with different arguments.
msgeval() defines the one-arg function to be returned, while _msgeval() is a two-arg function.
Probably I could use a lambda expression to mediate the arguments. But, IMHO, that wouldn't increase readability.

dirk-thomas replied:
Oh, I missed that. It can stay as it is then. Sorry for the noise.

dirk-thomas commented:
@ros-pull-request-builder

ros-pull-request-builder commented:
Test failed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/262/

dirk-thomas commented:
@ros-pull-request-builder

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/263/

ros-pull-request-builder commented:
Test failed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/265/

rhaschke commented:
I rebased the pull request to integrate the fixups into their corresponding commit.
Does this make the build check fail?

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/264/

dirk-thomas commented:
One job failed since the hash was not available anymore when you rebased and force pushed. The newly triggered one passed.

@esteve @tfoote @wjwwood Please review (and if you can give it a try to ensure we don't break anything).

dirk-thomas commented:
@ros-pull-request-builder

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/266/

wjwwood commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -280,50 +280,62 @@ def _rostopic_bw(topic, window_size=-1):
+    for f in fields:
+        if '[' in f:
+            field_name, rest = f.split('[', 1)
+            rest = rest[:-1]  # slice content, removing closing bracket
```
Is there an assumption here about having a trailing `]`? Should there be an assertion to that end? If it's not there, should there be an error? It seems to me that an input of `foo[123` could be incorrectly parsed the same as `foo[12]`. Maybe this isn't an issue due to the overall context, but it isn't clear to me here. There seemed to be a similar check and error on the removed lines 305-307.

wjwwood replied:
It looks like the test cases do not cover this type of "syntax" exception either.

rhaschke replied:
Valid point. I added an assertion and an appropriate unittest.

wjwwood commented:
Other than the comments this looks good to me, @rhaschke thanks a lot for the contribution and iterating with us on it.

rhaschke committed:
added consistency check for slicing brackets (as well as unittest)

missing closing bracket will be reported to stderr

rhaschke committed:
renamed _msgeval() by msgeval() and use lambda expression to mediate arguments

ros-pull-request-builder commented:
Test failed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/267/

dirk-thomas commented:
@ros-pull-request-builder

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/268/

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -280,50 +280,60 @@ def _rostopic_bw(topic, window_size=-1):
+    for f in fields:
+        if '[' in f:
+            field_name, rest = f.split('[', 1)
+            assert rest[-1] == ']', "missing closing ']' in slice spec '%s'" % f
```
Instead of using `assert` the function should report the problem in the same way as other similar errors (print error message to `stderr` and return `None`).

wjwwood replied:
I agree, that would be preferable.

rhaschke replied:
Sorry, it was too late yesterday. I simply took code from `_get_array_index_or_slice_object()` without noticing that the AssertionError was catched.
Replaced now assert by a simple if and an appropriate warning on stderr.

ros-pull-request-builder commented:
Test passed.
Refer to this link for build results: http://jenkins.ros.org/job/_pull_request-indigo-ros_comm/271/

dirk-thomas commented on tools/rostopic/src/rostopic/__init__.py:
```diff
@@ -788,7 +800,9 @@ def _rostopic_echo(topic, callback_echo, bag_file=None, echo_all_topics=False):
                     type_information = submsg_class._slot_types[index]
                     if fields:
-                        submsg_class = roslib.message.get_message_class(type_information)
+                        submsg_class = roslib.message.get_message_class(type_information.split('[',1)[0])
```
Please add a space between `,` and `1` following PEP8

dirk-thomas replied:
Never mind, I will just apply the change when squashing all commits. Thanks.

dirk-thomas commented:
Squashed and merged in 7022c7f53dac06587f2c46e185aa3d8286f9e575

Thank you for contributing this enhancement!

ros-pull-request-builder commented:
Can one of the admins verify this patch?

rhaschke commented:
Thanks for patiently guiding me through your code style rules.
